Add basic Qt wallet test

ryanofsky opened the pull request:
This adds a basic unit test for the Qt send coins dialog. I have more changes (for a future PR) that extend this to cover new Qt bumpfee code in #9697.

Previously, I brought up the topic of qt unit tests in https://github.com/bitcoin/bitcoin/pull/9817#issuecomment-281351497 and https://github.com/bitcoin/bitcoin/pull/9697#discussion_r102971669.

ryanofsky committed:
Make test_bitcoin.cpp compatible with Qt Test framework

Move Boost.Test main function and global overrides to a new test_bitcoin_main.cpp file.

ryanofsky committed:
Make qt test compatible with TestChain100Setup framework

Reset global state after rpc tests, and remove unnecessary ECC initialization
to prevent assert error if it is initialized twice.

ryanofsky committed:
Add simple qt wallet test sending a transaction

luke-jr commented:
I almost wonder if it might be useful to have a RPC wrapper for the GUI (possibly in addition to this) so we can just use the RPC tests with it as well.

fanquake commented:
Failed with:
```
==============================================
   Bitcoin Core 0.14.99: src/test-suite.log
==============================================
# TOTAL: 2
# PASS:  1
# SKIP:  0
# XFAIL: 0
# FAIL:  1
# XPASS: 0
# ERROR: 0
.. contents:: :depth: 2
FAIL: qt/test/test_bitcoin-qt
=============================
This application failed to start because it could not find or load the Qt platform plugin "xcb"
in "".
Reinstalling the application may fix this problem.
```

laanwj commented:
Concept ACK

MarcoFalke commented:
Concept ACK for sure, as we basically don't have any gui tests right now. In the long term I'd prefer a solution that runs on a higher abstraction level. Your current approach is pretty much mimicking functional tests on a low (unit test) level.

ryanofsky commented:
> I'd prefer a solution that runs on a higher abstraction level. Your current approach is pretty much mimicking functional tests on a low (unit test) level.

I don't think I understand. I'm not doing anything original here, just following the approach of http://doc.qt.io/qt-5/qttestlib-tutorial3-example.html.

Is it that instead of the writing a test that sends qt events, you would prefer to see a test that starts an x server and sends keyboard and mouse events?

jonasschnelli commented:
I really like this.
Concept ACK.

MarcoFalke commented:
> Is it that instead of the writing a test that sends qt events, you would
prefer to see a test that starts an x server and sends keyboard and mouse
events?

Indeed. I was dreaming of a gui test framework where you just press a
"record this" button. Then records discrete actions (e.g. "Press: Receive
tab", "Press: Copy address", other gui: "Press: Send tab", "Insert:
$address", ...) instead of the exact mouse and keyboard movements, such
that it works on different screen resolutions. Though, we also need a way
to wait for the gui to catch up on longer tasks such as importmulti with
rescan, so I don't think this is a weekend project.





On Sat, Mar 11, 2017 at 12:36 PM, Jonas Schnelli <notifications@github.com>
wrote:

> I really like this.
> Concept ACK.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/9974#issuecomment-285861292>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AGGmv2gRizL47AbknwIa11DdwdLFQFBrks5rkoc4gaJpZM4MZ6Pd>
> .
>

fanquake commented on src/qt/test/test_main.cpp:
```diff
@@ -46,7 +45,7 @@ int main(int argc, char *argv[])
 
     // Don't remove this, it's needed to access
     // QCoreApplication:: in the tests
```
Comment needs removing/updating?

ryanofsky replied:
It's still accurate, but I expanded the comment in f4d28fbe1b8041ce358d80ecb0e1fa074eee359c.

fanquake commented:
Still failing Travis with the same issue, @theuni should probably take a look here.

MarcoFalke commented on src/qt/test/test_main.cpp:
```diff
@@ -65,7 +64,11 @@ int main(int argc, char *argv[])
         fInvalid = true;
+#ifdef ENABLE_WALLET
+    WalletTests test5;
+    if (QTest::qExec(&test5) != 0)
```
style nit: We use braces now; Always.

ryanofsky replied:
Added in b61b34c89dbe3946e21147dcfe620a4a6ff536ed

ryanofsky committed:
Add new test_bitcoin-qt static library dependencies

Avoids following error when qt is statically linked into the test binary, as on
travis:

This application failed to start because it could not find or load the Qt platform plugin "xcb"
in "".

ryanofsky committed:
Enable xvfb in travis to allow running test_bitcoin-qt

Avoids following error:

QXcbConnection: Could not connect to display

ryanofsky committed:
Add braces to if statements in Qt test_main

ryanofsky reviewed (commented):
Finally got test to pass on travis.

jonasschnelli commented:
A lock is missing.
```
diff --git a/src/qt/test/wallettests.cpp b/src/qt/test/wallettests.cpp
index 81b358e..d63f4ec 100644
--- a/src/qt/test/wallettests.cpp
+++ b/src/qt/test/wallettests.cpp
@@ -73,6 +73,7 @@ void WalletTests::walletTests()
     test.CreateAndProcessBlock({}, GetScriptForRawPubKey(test.coinbaseKey.GetPubKey()));
     bitdb.MakeMock();
     CWallet wallet("wallet_test.dat");
+    LOCK(wallet.cs_wallet);
     bool firstRun;
     wallet.LoadWallet(firstRun);
     wallet.SetAddressBook(test.coinbaseKey.GetPubKey().GetID(), "", "receive");
```
How could this pass travis with the missing lock?

Will properly ACK after apply the diff above.

Question: is it possible to create a screenshot (or a bitmap/JPG/PNG) of the UI elements you open?
A such form of an UI export would be great. We could add this then to a CI and have plenty of screenshots from various operating systems in order to verify the UI correctness.

ryanofsky commented:
> A lock is missing.

Will add. I'm not sure the way the lock is acquired in the diff is exactly right, because cs_main is supported to be acquired before cs_wallet, and some wallet methods acquire both.

> How could this pass travis with the missing lock?

Locally, the test works fine for me with the missing lock. On travis, every configuration we have either specifies NO_QT or NO_WALLET, or omits RUN_TESTS, so neither PaymentServerTests code nor this new WalletTests code currently executes.

> Question: is it possible to create a screenshot (or a bitmap/JPG/PNG) of the UI elements you open?
A such form of an UI export would be great. We could add this then to a CI and have plenty of screenshots from various operating systems in order to verify the UI correctness.

Something like this should be possible. To help debug this change, I frequently added `widget.show(); QEventLoop().exec();` calls to be able to see and interact with individual test widgets, so it should be possible to capture images also (maybe with [QWidget::grab](http://doc.qt.io/qt-5/qwidget.html#grab)?)

ryanofsky commented:
Fixed locking bug in 739f19ca969c2ae351e18d2bbb424237830f54f1

Squashed 739f19ca969c2ae351e18d2bbb424237830f54f1 -> 9576b015a107b98fc950c574ed01d993b388d7c9 ([pr/qt-test.7](https://github.com/ryanofsky/bitcoin/commits/pr/qt-test.7) -> [pr/qt-test.8](https://github.com/ryanofsky/bitcoin/commits/pr/qt-test.8))

MarcoFalke commented:
utACK 9576b015a107b98fc950c574ed01d993b388d7c9

jonasschnelli commented:
Tested ACK 9576b015a107b98fc950c574ed01d993b388d7c9

jnewbery commented:
This is great. Hopefully I'll have coverage running soon so we can see how much extra code you've managed to exercise.

Post-merge tested ACK https://github.com/bitcoin/bitcoin/commit/9576b015a107b98fc950c574ed01d993b388d7c9.

paveljanik commented:
BTW, compiling this is not possible with gcc-4.8.5, the default gcc on openSUSE 42.2:

```
/usr/include/boost/signals2/connection.hpp: In function 'uint256 {anonymous}::SendCoins(CWallet&, SendCoinsDialog&, const CBitcoinAddress&, CAmount)':
/usr/include/boost/signals2/connection.hpp:234:7: error: 'boost::signals2::scoped_connection::scoped_connection(const boost::signals2::scoped_connection&)' is private
       scoped_connection(const scoped_connection &other);
       ^
qt/test/wallettests.cpp:47:6: error: within this context
     });
      ^
Makefile:7991: recipe for target 'qt/test/qt_test_test_bitcoin_qt-wallettests.o' failed
make[2]: *** [qt/test/qt_test_test_bitcoin_qt-wallettests.o] Error 1
```

ryanofsky commented:
Possible fix might be to change:

```
boost::signals2::scoped_connection c = wallet.NotifyTransactionChanged.connect(...);
```

to:

```
boost::signals2::scoped_connection c(wallet.NotifyTransactionChanged.connect(...));
```

paveljanik commented:
Hmm, all gcc's on openSUSE 42.2, ie. 4.8.5, 5.3.1 and 6.2.1, fail on the current master...

paveljanik commented:
With https://github.com/bitcoin/bitcoin/pull/9974#issuecomment-287550867 change, the build is OK. Will you please PR it?

paveljanik commented:
And another issue here. On plain Debian Jessie, with the gcc version 4.9.2 (Debian 4.9.2-10), this doesn't compile with:

```
qt/test/wallettests.cpp:34:6: error: no matching function for call to 'QTimer::singleShot(int, Qt::TimerType, {anonymous}::ConfirmSend()::<lambda()>)'
     });
      ^
qt/test/wallettests.cpp:34:6: note: candidates are:
In file included from /usr/include/x86_64-linux-gnu/qt5/QtCore/QTimer:1:0,
                 from ./qt/sendcoinsdialog.h:13,
                 from qt/test/wallettests.cpp:7:
/usr/include/x86_64-linux-gnu/qt5/QtCore/qtimer.h:81:17: note: static void QTimer::singleShot(int, const QObject*, const char*)
     static void singleShot(int msec, const QObject *receiver, const char *member);
                 ^
/usr/include/x86_64-linux-gnu/qt5/QtCore/qtimer.h:81:17: note:   no known conversion for argument 2 from 'Qt::TimerType' to 'const QObject*'
/usr/include/x86_64-linux-gnu/qt5/QtCore/qtimer.h:82:17: note: static void QTimer::singleShot(int, Qt::TimerType, const QObject*, const char*)
     static void singleShot(int msec, Qt::TimerType timerType, const QObject *receiver, const char *member);
                 ^
/usr/include/x86_64-linux-gnu/qt5/QtCore/qtimer.h:82:17: note:   candidate expects 4 arguments, 3 provided
Makefile:7991: recipe for target 'qt/test/qt_test_test_bitcoin_qt-wallettests.o' failed
make[2]: *** [qt/test/qt_test_test_bitcoin_qt-wallettests.o] Error 1
```
